jupyterhub/templates/hub/netpol.yaml

```diff
@@ -13,17 +13,39 @@ spec:
     - Ingress
     - Egress
 
+  # IMPORTANT:
+  # NetworkPolicy's ingress "from" and egress "to" rule specifications require
+  # great attention to detail. A quick summary is:
+  #
+  # 1. You can provide "from"/"to" rules that provide access either ports or a
+  #    subset of ports.
+  # 2. You can for each "form"/"to" rule provide any number of
+  #    "sources"/"destinations" of four different kinds.
+  #    - podSelector                        - targets pods with a certain label in the NetworkPolicy's namespace
+  #    - namespaceSelector                  - targets all pods within namespaces with a certain label
+  #    - namespaceSelector and podSelector  - targets pods with a certain label within namespaces with a certain label
+  #    - ipBlock                            - targets network traffic from/to a set of IP address ranges
+  #
+  # Read more at: https://kubernetes.io/docs/concepts/services-networking/network-policies/#behavior-of-to-and-from-selectors
+  #
   ingress:
     # allowed pods (hub.jupyter.org/network-access-hub) --> hub
     - ports:
         - port: http
       from:
+        # source 1 - labelled pods
         - podSelector:
             matchLabels:
               hub.jupyter.org/network-access-hub: "true"
+          namespaceSelector:
+            matchLabels: {}   # without this, the podSelector would only consider pods in the local namespace
+        # source 2 - pods in labelled namespaces
+        - namespaceSelector:
+            matchLabels:
+              hub.jupyter.org/network-access-hub: "true"
```

Comment:
This is the key part to review, which is repeated four times across the PR.

Reply:
What do you think about moving source 2 to values.yaml? Currently `networkPolicy.ingress: []`, if source 2 is removed from here and `networkPolicy.ingress` is changed to:
```yaml
hub:
  networkPolicy:
    ingress:
      - namespaceSelector:
            matchLabels:
              hub.jupyter.org/network-access-hub: "true"
```
This makes it easy to override, either to allow all traffic without labels, or to block this traffic restoring the original behaviour?

Reply:
On second thoughts maybe I meant:
```
      from:
        # source 1 - labelled pods
        - podSelector:
            matchLabels:
              hub.jupyter.org/network-access-hub: "true"
```
_Without_ the namespaceSelector:

then in Values.yaml
```yaml
hub:
  networkPolicy:
    ingress:
      - namespaceSelector:
            matchLabels:
              hub.jupyter.org/network-access-hub: "true"
            namespaceSelector:
              matchLabels: {}
      - namespaceSelector:
          matchLabels:
            hub.jupyter.org/network-access-hub: "true"
```

In other words `netpol.yaml` contains the rules that are strictly necessary for JupyterHub to function, then Values.yaml contains the rules that will work for most people but with enough flexibility for them to be relaxed or restricted.

Reply:
```yaml
hub:
  networkPolicy:
    ingress:
      - ports:
        - port: http
        from:
        - podSelector:
            matchLabels:
              hub.jupyter.org/network-access-hub: "true"
            namespaceSelector:
              matchLabels: {}
        - namespaceSelector:
            matchLabels:
              hub.jupyter.org/network-access-hub: "true"
```

This would be the equivalent thing I think. Hmm... I think this is just massively complicated, and I think the maintenance complexity and such for anyone making anything custom will be too much for what we can manage.

I'd say we have it as I suggest, and if strictly required, we have a toggle to influence that default to not accept inter-namespace labeling.

Reply:
Imagine for example if a user wants to add a rule, without changing the defaults - since hub.networkPolicy.ingress is a list, it will be replaced by the user which in my mind likely will be a not-intentional action.

I think it may be only you and me who will be able to help when users run into issues stemming from such configuration changes, because I doubt anyone else will be able to do that without significant code inspection - which you and me have right now from working on this. I think we are stretched thin as it is to maintain z2jh, kubespawner, binderhub, and the k8s related jupyterhub repos.

Reply:
Maybe if we work through one or more examples that'll help to see what the impact is on users?

E.g. On clusters where namespaces are used to segregate users:
- I don't have admin on other namespaces
- I don't want ingress from other namespaces unless I say so
- I'd like to enable ingress from the namespace `nginx-ingress` where there's a cluster level ingress controller

What would my Z2JH config be?

Feel free to add the other examples you have in mind.

Reply:
Expanding further on my point, my understanding of this PR is that if you have cluster admin rights there should be no need to modify the default ingress policy since the recommendation is to add labels to your pods? Can you give some examples of ingress rules that you can see being required?

Reply:
Expanding further on my point, my understanding of this PR is that if you have cluster admin rights there should be no need to modify the default ingress policy since the recommendation is to add labels to your pods? Can you give some examples of ingress rules that you can see being required?

Reply:
@manics deal!

## Overview of what we configure ingress for

We have networkPolicy resources declaring allowed ingress from four locations, for three pods.
- proxy's API port
- proxy's HTTP(S) port(s)
- hub's HTTP port
- singleuser-server's HTTP port

There is already some predefined escalation of access by going through the proxy's HTTP(s) port(s) though, which typically are publicly exposed on the internet, as you can it proxies to the singleuser-server's HTTP port through it (/user/myusername) and the hub's port through it (/hub). The proxy's API port should probably not be publicly exposed, or perhaps ever used except by JupyterHub.

## Ability to configure zero trust
Assuming we have no trust for other namespaces, and think they maliciously want to sent crap to our pods in the JupyterHub namespace, then we can make it configurable to disallow that.

The "risky" configuration discussed is that we allow a) under "source 1" pods with the right labels in other namespaces with to access specific pods in our namespace, and b) pods in namespaces the right labels.

### Source 1 lockdown
We can lock down source 1 in two steps compared to this PR. Source 1 is now is a `podSelector and namespaceSelector`, and we can make it just become a `podSelector`, that is the most tight lock down. The second tightest lockdown is to require that the namespace is also correctly labelled.

Note that we cannot specify individual namespaces, only select them based on labels.

### Source 2 lockdown
We can lock down source 2 only with a single step. Source 2 is now a `namespaceSelector`, and we can make it not be added with a boolean flag per networkPolicy resource

### Source 1 and 2 configuration states
We get 3x2 permutations with some nonsensical combinations.

Permutation # | Source 1 configuration | Source 2 configuration | Note
-|-|-|-
1 | Local only | Disabled | __No inter-namespace ingress allowed.__
2 | Namespace labelled | Disabled | __Inter-namespace ingress allowed if both namespace and pod is labelled.__
3 | All namespaces | Disabled | __Practically nonsensical__. To only allow Pod labels, but not namespace labels, is probably to allow a lower set of permissions I think?
4 | Local only | Enabled | __Practically nonsensical__ - This just limit the configuration resolution in comparison to other options.
5 | Namespace labelled | Enabled | __Theoretically nonsensical__ - Source 1 will never be needed.
6 | All namespaces | Enabled | __Inter-namespace ingress is allowed if either namespace or pod is labelled.__

There are in other words three reasonable states:
1. No inter-namespace ingress allowed.
2. Inter-namespace ingress allowed if both namespace and pod is labelled.
3. Inter-namespace ingress is allowed if either namespace or pod is labelled.

### Conclusion
We would provide a configuration that has three states for our four different ingress sources and associated labels.

```yaml
hub|singleuser|proxy:
  networkPolicy:
    interNamespaceIngressForHTTPPort: deny | requireBothPodAndNamespaceLabel | requireEitherPodOrNamespaceLabel
    # for proxy.networkPolicy, we would also have...
    interNamespaceIngressForAPIPort: deny | requireBothPodAndNamespaceLabel | requireEitherPodOrNamespaceLabel
```

Reply:
I guess permutation 4 is potentially relevant, because it may be deemed safer to allow a namespace label than a pod label, and one may not want care about ending up allowing all pods in that namespace.

```yaml
hub|singleuser|proxy:
  networkPolicy:
    interNamespaceIngressForHTTPPort: deny | requireBothPodAndNamespaceLabel | requireEitherPodOrNamespaceLabel | requireNamespaceLabel
    # for proxy.networkPolicy, we would also have...
    interNamespaceIngressForAPIPort: deny | requireBothPodAndNamespaceLabel | requireEitherPodOrNamespaceLabel | requireNamespaceLabel
```

Reply:
Thanks for explaining, I now understand what your thinking is!
Is the latest proposal effectively three ingress network policies, two controlled by booleans and one `<anything>`?
```yaml
hub|singleuser|proxy:
  networkPolicy:
    interNamespaceIngressForHTTPPort: deny | requireBothPodAndNamespaceLabel | requireEitherPodOrNamespaceLabel | requireNamespaceLabel
    interNamespaceIngressForAPIPort: deny | requireBothPodAndNamespaceLabel | requireEitherPodOrNamespaceLabel | requireNamespaceLabel
    ingress: []
```
If so it's not my preferred option but I can live with it :smile: 

An alternative to `interNamespaceIngressForAPIPort` is to add hub-api to the list of components: `hub|hubapi|singleuser|proxy` so you just have `interNamespaceIngress`, dunno if that's better or worse though.

My other concern was whether we're being "good citizens" in the helm/k8s ecosystem by imposing the use of labels this way? It's fine to say Z2JH is aimed at people with beginner to intermediate experience of K8S, and that we recommend people use labels because we think this is the easiest solution (and if you're an advanced user these are the instructions to do your own thing....), but for me it'd be good to be explicit about that. Getting off topic, but stating this will also help evaluate feature requests- do we add an easy to configure option, or do we add an override and say "override this here, but at your own risk"

Reply:
> Is the latest proposal effectively three ingress network policies, two controlled by booleans and one <anything>?

Hmmmm, I don't understand... But I meant....

```yaml
# ingress / egress are additional rules that can permit
# other kinds of traffic other than what the default configured rules will provide

# Allowed options for interNameSpaceIngressForXXXXPort are:
# - deny
# - requireBothPodAndNamespaceLabel
# - requireEitherPodOrNamespaceLabel
# - requireNamespaceLabel

# Default values suggested below
hub:
  networkPolicy:
    interNamespaceIngressForHTTPPort: deny 
    ingress: []
    egress:
      - to:
          - ipBlock:
              cidr: 0.0.0.0/0

proxy:
  networkPolicy:
    interNamespaceIngressForHTTPPort: requireEitherPodOrNamespaceLabel
    interNamespaceIngressForAPIPort: deny
    ingress: []
    egress:
      - to:
          - ipBlock:
              cidr: 0.0.0.0/0

singleuser:
  networkPolicy:
    interNamespaceIngressForHTTPPort: deny
    ingress: []
    egress:
      - to:
          - ipBlock:
              cidr: 0.0.0.0/0
              except:
                - 169.254.169.254/32
```

> If so it's not my preferred option but I can live with it smile

Hehe, pweh! I would not be able to live with the idea of users copy-pasting complicated configuration which end up becoming outdated minimal requirements in the future etc, or just the fact that it becomes extremely hard to understand how something has gone wrong if that becomes part of the bug reports - then everyone need to be able to perfectly spot syntax errors in that configuration to rule out its an issue, and even those who can will loose time on ruling it out as the cause of the issue reported.

> My other concern was whether we're being "good citizens" in the helm/k8s ecosystem by imposing the use of labels this way?

We are imposing them as part of secure defaults, which can be disabled. If you use network policy resources, that actually has some meaning by locking thigns down, then in my mind, we are not imposing anything other than what is stemming from secure practices part of k8s - they have chosen the NetworkPolicy resource API where labels of pods/namespaces have impact.

An option is to allow ingress from all namespaces without any label to the proxy pod specifically by default, this would align with the 0.9.0/0.9.1 configuration, which I consider a bug as it made the label we had pointless.

Ah, okay, so we have a final option in practice: `allowAllIngress: false`.

> do we add an easy to configure option, or do we add an override and say "override this here, but at your own risk"

I feel that there has been a bit of requests like this, and even though people get to use things at their own risk, it typically end up being problematic for maintainers in the end anyhow. Example, someone overrides the hub image, then they report an issue, and suddenly the issue's cause can stem from another source that it couldn't if the user had not had the option to override the hub image. I have spent several hours debugging these kinds of issues to end up concluding it was because this. Due to that experience, I don't think it helps that much to add a "use at your own risk" note.

Reply:
>An alternative to interNamespaceIngressForAPIPort is to add hub-api to the list of components: hub|hubapi|singleuser|proxy so you just have interNamespaceIngress, dunno if that's better or worse though.

@manics The hub doesn't expose its API on a dedicated port, but the proxy server does. It has a REST API to get configured by JupyterHub in how it should route traffic on its port that proxies traffic to hub/users/reigstered services